Emcee and one other fix

amn3142 opened the pull request:
Emcee optimizer in centroiding and one other fix in the base psf masking so that it works when all a pixel is always zero in all stars (e.g. near edges).

sibirrer commented on psfr/psfr.py:
```diff
@@ -403,6 +413,48 @@ def centroid_fit(data, model, mask=None, variance=None, oversampling=1):
         required shift in units of pixels in the data such that the model matches best the data
     """
 
+    def getProbability(pars, variance, mask, oversampling, model, data):
```
this definition is basically the same (modulo a minus sign) fo the _minimise() function that is already implemented, so no need to duplicate the code here

sibirrer commented on psfr/psfr.py:
```diff
@@ -377,7 +387,7 @@ def _linear_amplitude(data, model, variance=None, mask=None):
 
 
-def centroid_fit(data, model, mask=None, variance=None, oversampling=1):
+def centroid_fit(data, model, mask=None, variance=None, oversampling=1,optimizer_type = 'NM'):
```
the new input variable is not documented (see documentation strings below).

sibirrer commented on psfr/psfr.py:
```diff
@@ -419,9 +471,16 @@ def _minimize(x, data, model, variance=None, mask=None, oversampling=1):
-    return x.x
+
+
+    if optimizer_type == 'NM':
```
I would give this option a more meaningful name 'NM' is not self-explanatory. Perhaps set string to 'Nelder-Mead'?

sibirrer commented on psfr/psfr.py:
```diff
@@ -419,9 +471,16 @@ def _minimize(x, data, model, variance=None, mask=None, oversampling=1):
+        x = scipy.optimize.minimize(_minimize, init, args=(data, model, variance, mask, oversampling),
+                                bounds=((-1.5, 1.5), (-1.5, 1.5)), method='Nelder-Mead')
+        return x.x
+    if optimizer_type =='emcee':
```
change logic to 'else if'

sibirrer commented on psfr/psfr.py:
```diff
@@ -419,9 +471,16 @@ def _minimize(x, data, model, variance=None, mask=None, oversampling=1):
+        return x.x
+    if optimizer_type =='emcee':
+        x =_minimize_emcee(init,data,model,variance, mask,oversampling, nwalkers = 10,niter= 100,startStep = 0.1)
+
```
add 'else:' statement and then give a raise statement:
else:
 raise ValueError('optimization type %s is not supported. Please use xxx instead.' % optimizer_type)

sibirrer commented on psfr/psfr.py:
```diff
@@ -454,15 +513,19 @@ def combine_psf(kernel_list_new, kernel_old, mask_list=None, weight_list=None, f
     for j, kernel_new in enumerate(kernel_list_new):
+
         for k in range(symmetry):
+
```
remove spacing here

sibirrer commented on psfr/psfr.py:
```diff
@@ -493,8 +556,11 @@ def combine_psf(kernel_list_new, kernel_old, mask_list=None, weight_list=None, f
     elif stacking_option == 'mean':
         kernel_new = np.average(kernel_list, weights=weights, axis=0)
+
+
```
remove double spacing, this is against PEP8 standard

sibirrer commented on psfr/psfr.py:
```diff
@@ -403,6 +413,48 @@ def centroid_fit(data, model, mask=None, variance=None, oversampling=1):
+
+    def _minimize_emcee(x, data, model, variance, mask, oversampling, nwalkers=25, niter=100, startStep=0.1,
+                        nrestart=2):
+        import emcee
```
emcee does not appear in the requirements of the package

sibirrer replied:
I would actually rather use the PSO. You can use the one implemented in lenstronomy (then no other dependancies are needed). You find it e.g. here https://github.com/lenstronomy/lenstronomy/blob/11e2733cbabd5b8565adf31a92a135b94a19cee6/lenstronomy/Sampling/sampler.py#L103

There is also no need to re-start the sampler. I think it's going to be much more efficient

sibirrer reviewed (changes requested):
Congrats for your first pull request!
I gave quite some comments, mostly minor. I suggest you are using/implementing the PSO instead of the emcee.
Also, the tests are failing in this PR, so something is not working properly. You find the error message in the thread and link to GitHub actions

Let me know if you have questions and thanks a lot!

amn3142 commented:
Ok! Thanks Simon, sorry for the mess. What is the appropriate way to proceed? Is there a pullback, or do I just go off of your feedback?



On Sep 10, 2022, at 9:00 PM, Simon Birrer ***@***.******@***.***>> wrote:



@sibirrer requested changes on this pull request.

Congrats for your first pull request!
I gave quite some comments, mostly minor. I suggest you are using/implementing the PSO instead of the emcee.
Also, the tests are failing in this PR, so something is not working properly. You find the error message in the thread and link to GitHub actions

Let me know if you have questions and thanks a lot!

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751497>:

> @@ -403,6 +413,48 @@ def centroid_fit(data, model, mask=None, variance=None, oversampling=1):
         required shift in units of pixels in the data such that the model matches best the data
     """

+    def getProbability(pars, variance, mask, oversampling, model, data):


this definition is basically the same (modulo a minus sign) fo the _minimise() function that is already implemented, so no need to duplicate the code here

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751527>:

> @@ -377,7 +387,7 @@ def _linear_amplitude(data, model, variance=None, mask=None):
     return amp


-def centroid_fit(data, model, mask=None, variance=None, oversampling=1):
+def centroid_fit(data, model, mask=None, variance=None, oversampling=1,optimizer_type = 'NM'):


the new input variable is not documented (see documentation strings below).

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751695>:

> @@ -419,9 +471,16 @@ def _minimize(x, data, model, variance=None, mask=None, oversampling=1):
         return chi2

     init = np.array([0, 0])
-    x = scipy.optimize.minimize(_minimize, init, args=(data, model, variance, mask, oversampling),
-                                bounds=((-5, 5), (-5, 5)), method='Nelder-Mead')
-    return x.x
+
+
+    if optimizer_type == 'NM':


I would give this option a more meaningful name 'NM' is not self-explanatory. Perhaps set string to 'Nelder-Mead'?

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751734>:

> @@ -419,9 +471,16 @@ def _minimize(x, data, model, variance=None, mask=None, oversampling=1):
         return chi2

     init = np.array([0, 0])
-    x = scipy.optimize.minimize(_minimize, init, args=(data, model, variance, mask, oversampling),
-                                bounds=((-5, 5), (-5, 5)), method='Nelder-Mead')
-    return x.x
+
+
+    if optimizer_type == 'NM':
+        x = scipy.optimize.minimize(_minimize, init, args=(data, model, variance, mask, oversampling),
+                                bounds=((-1.5, 1.5), (-1.5, 1.5)), method='Nelder-Mead')
+        return x.x
+    if optimizer_type =='emcee':


change logic to 'else if'

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751836>:

> @@ -419,9 +471,16 @@ def _minimize(x, data, model, variance=None, mask=None, oversampling=1):
         return chi2

     init = np.array([0, 0])
-    x = scipy.optimize.minimize(_minimize, init, args=(data, model, variance, mask, oversampling),
-                                bounds=((-5, 5), (-5, 5)), method='Nelder-Mead')
-    return x.x
+
+
+    if optimizer_type == 'NM':
+        x = scipy.optimize.minimize(_minimize, init, args=(data, model, variance, mask, oversampling),
+                                bounds=((-1.5, 1.5), (-1.5, 1.5)), method='Nelder-Mead')
+        return x.x
+    if optimizer_type =='emcee':
+        x =_minimize_emcee(init,data,model,variance, mask,oversampling, nwalkers = 10,niter= 100,startStep = 0.1)
+


add 'else:' statement and then give a raise statement:
else:
raise ValueError('optimization type %s is not supported. Please use xxx instead.' % optimizer_type)

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751846>:

>          for k in range(symmetry):
+


remove spacing here

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751890>:

>      elif stacking_option == 'mean':
         kernel_new = np.average(kernel_list, weights=weights, axis=0)
+
+


remove double spacing, this is against PEP8 standard

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967751958>:

> +        model_shifted = shift_psf(psf_center=model, oversampling=oversampling, shift=pars, degrade=True,
+                                  n_pix_star=len(data))
+
+        amp = _linear_amplitude(data, model_shifted, variance=variance, mask=mask)
+
+        chi2 = -1 * np.sum((data - model_shifted * amp) ** 2 / variance * mask)
+        return chi2
+
+
+    ###this function is better if the initial guess (0,0) is far from the center of the pixel.
+    # ##The starting step valueis optimized to my particular pixel size/PSF width ratio,
+    # ##so we probably want someone to be ablte to pass a variable here for more general use.
+
+    def _minimize_emcee(x, data, model, variance, mask, oversampling, nwalkers=25, niter=100, startStep=0.1,
+                        nrestart=2):
+        import emcee


emcee does not appear in the requirements of the package

________________________________

In psfr/psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r967752144>:

> +        model_shifted = shift_psf(psf_center=model, oversampling=oversampling, shift=pars, degrade=True,
+                                  n_pix_star=len(data))
+
+        amp = _linear_amplitude(data, model_shifted, variance=variance, mask=mask)
+
+        chi2 = -1 * np.sum((data - model_shifted * amp) ** 2 / variance * mask)
+        return chi2
+
+
+    ###this function is better if the initial guess (0,0) is far from the center of the pixel.
+    # ##The starting step valueis optimized to my particular pixel size/PSF width ratio,
+    # ##so we probably want someone to be ablte to pass a variable here for more general use.
+
+    def _minimize_emcee(x, data, model, variance, mask, oversampling, nwalkers=25, niter=100, startStep=0.1,
+                        nrestart=2):
+        import emcee


I would actually rather use the PSO. You can use the one implemented in lenstronomy (then no other dependancies are needed). You find it e.g. here https://github.com/lenstronomy/lenstronomy/blob/11e2733cbabd5b8565adf31a92a135b94a19cee6/lenstronomy/Sampling/sampler.py#L103

There is also no need to re-start the sampler. I think it's going to be much more efficient

—
Reply to this email directly, view it on GitHub<https://github.com/sibirrer/psfr/pull/7#pullrequestreview-1103231632>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEA6XOVZ433FBCZE22GVUSLV5VKMVANCNFSM6AAAAAAQJD3NSI>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

--------------------------------------------
Anna M. Nierenberg

Assistant Professor
Department of Physics, UC Merced
annanierenberg.com<http://annanierenberg.com>

sibirrer commented:
@amn3142 you should go off my feedback, but also can say *no* :) if you make your next commit and push it to your GitHub, it automatically updates this pull request. I would reply to individual comments I made to focus the discussion around things.

sibirrer commented:
> Hi Simon, So the code is passing the centroiding test as is on my computer. One thing I noticed is that in PSFr, the bounds on the Nelder Meade optimizer are : bounds=((-1.5, 1.5), (-1.5, 1.5)) (beyond the range of the centroid offset in the model). When I run the centroiding I get an error message: python test_centroid.py /home/anierenberg/repositories/miniconda3/envs/py3/lib/python3.6/site-packages/scipy/optimize/_minimize.py:534: RuntimeWarning: Method Nelder-Mead cannot handle constraints nor bounds. However, the optimization completes and I get the result: [-3.4999921 2.20119529] I wonder if the bounds are being obeyed in the test implementation for some reason? (different python version?) ANna 

Hi Anna, it might be that you have an older scipy version installed that does not support it. Also, if you want to change the bounds in the optimizer, feel free. The current bounds assume that you are well centroided within a pixel but probably does not change anything for the current cases.

amn3142 commented:
I think this is causing the test to fail because the true value in the test is outside of this.

On Sep 11, 2022, at 9:12 PM, Simon Birrer ***@***.***> wrote:

﻿

Hi Simon, So the code is passing the centroiding test as is on my computer. One thing I noticed is that in PSFr, the bounds on the Nelder Meade optimizer are : bounds=((-1.5, 1.5), (-1.5, 1.5)) (beyond the range of the centroid offset in the model). When I run the centroiding I get an error message: python test_centroid.py /home/anierenberg/repositories/miniconda3/envs/py3/lib/python3.6/site-packages/scipy/optimize/_minimize.py:534: RuntimeWarning: Method Nelder-Mead cannot handle constraints nor bounds. However, the optimization completes and I get the result: [-3.4999921 2.20119529] I wonder if the bounds are being obeyed in the test implementation for some reason? (different python version?) ANna

Hi Anna, it might be that you have an older scipy version installed that does not support it. Also, if you want to change the bounds in the optimizer, feel free. The current bounds assume that you are well centroided within a pixel but probably does not change anything for the current cases.

—
Reply to this email directly, view it on GitHub<https://github.com/sibirrer/psfr/pull/7#issuecomment-1243196021>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEA6XOVDVGXIKQPGV6L6YTLV52UUHANCNFSM6AAAAAAQJD3NSI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

amn3142 commented:
Just to confirm you are happy if the psfr code uses a different convention/implementation of PSO, and not the particular one currently in sampler.py (because the PSO implementation in sampler.py  requires the creation of a LikelihoodModule object which is designed for the lensing case.)





On Sep 11, 2022, at 9:22 PM, Anna Nierenberg ***@***.******@***.***>> wrote:

I think this is causing the test to fail because the true value in the test is outside of this.

On Sep 11, 2022, at 9:12 PM, Simon Birrer ***@***.******@***.***>> wrote:

﻿


Hi Simon, So the code is passing the centroiding test as is on my computer. One thing I noticed is that in PSFr, the bounds on the Nelder Meade optimizer are : bounds=((-1.5, 1.5), (-1.5, 1.5)) (beyond the range of the centroid offset in the model). When I run the centroiding I get an error message: python test_centroid.py /home/anierenberg/repositories/miniconda3/envs/py3/lib/python3.6/site-packages/scipy/optimize/_minimize.py:534: RuntimeWarning: Method Nelder-Mead cannot handle constraints nor bounds. However, the optimization completes and I get the result: [-3.4999921 2.20119529] I wonder if the bounds are being obeyed in the test implementation for some reason? (different python version?) ANna

Hi Anna, it might be that you have an older scipy version installed that does not support it. Also, if you want to change the bounds in the optimizer, feel free. The current bounds assume that you are well centroided within a pixel but probably does not change anything for the current cases.

—
Reply to this email directly, view it on GitHub<https://github.com/sibirrer/psfr/pull/7#issuecomment-1243196021>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEA6XOVDVGXIKQPGV6L6YTLV52UUHANCNFSM6AAAAAAQJD3NSI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

--------------------------------------------
Anna M. Nierenberg

Assistant Professor
Department of Physics, UC Merced
annanierenberg.com<http://annanierenberg.com>

sibirrer commented:
> Just to confirm you are happy if the psfr code uses a different convention/implementation of PSO, and not the particular one currently in sampler.py (because the PSO implementation in sampler.py requires the creation of a LikelihoodModule object which is designed for the lensing case.) 

Yes indeed, I would import the PSO class from this file: https://github.com/lenstronomy/lenstronomy/blob/main/lenstronomy/Sampling/Samplers/pso.py
and then use it outside of the LikelihoodModule indeed. Then you can use it very similar to how you currently use the emcee.

amn3142 commented:
Ok I’ve implemented the PSO and addressed the other comments. I’ve also verified that it (and the original Nelder-Mead) now work (much faster than emcee as promised). Do I need to do a new pull request?

Thanks,

Anna

On Sep 12, 2022, at 2:37 PM, Simon Birrer ***@***.******@***.***>> wrote:



Just to confirm you are happy if the psfr code uses a different convention/implementation of PSO, and not the particular one currently in sampler.py (because the PSO implementation in sampler.py requires the creation of a LikelihoodModule object which is designed for the lensing case.)

Yes indeed, I would import the PSO class from this file: https://github.com/lenstronomy/lenstronomy/blob/main/lenstronomy/Sampling/Samplers/pso.py
and then use it outside of the LikelihoodModule indeed. Then you can use it very similar to how you currently use the emcee.

—
Reply to this email directly, view it on GitHub<https://github.com/sibirrer/psfr/pull/7#issuecomment-1244524409>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEA6XOUJBPIXG7JPZYVB5ELV56PBTANCNFSM6AAAAAAQJD3NSI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

--------------------------------------------
Anna M. Nierenberg

Assistant Professor
Department of Physics, UC Merced
annanierenberg.com<http://annanierenberg.com>

sibirrer commented on psfr/psfr.py:
```diff
@@ -325,10 +333,16 @@ def base_stacking(star_list, mask_list):
         weight_map += mask_list[i] * np.sum(star)
+
+    ###code can't handle situations where there is never a non-zero pixel
+    weight_map[weight_map == 0] = 1e-6
```
1e-6 might a bit too high depending on some applications. What about 1e-12? This is still in the range of floats to be handled numerically stable

amn3142 replied:
Just saw this, I'll change it.

sibirrer reviewed (commented):
This looks all great!
I am waiting for the continuous integration testing and the test coverage report (there are perhaps certain aspects that are not yet tested) Otherwise looks good. Thank you!

sibirrer reviewed (changes requested):
Tests are passing, and very fast!
Can you add a test case using the PSO instead of the Nelder-Mead algorithm?
You can have a look in the test folder and perhaps copy-past one example and instead are using the PSO option

sibirrer commented on tests/test_psfr.py:
```diff
@@ -280,3 +305,9 @@ def test_combine_psf():
+
+
+
+test_fit_centroid_pso()
```
you don't need to execute the definition here. py.test does it anyways :)
This is my last request before approving :)

sibirrer reviewed (commented):
Looks awesome!
One very minor last comment and then we can merge. Thank you very much!

amn3142 commented:
Ah shoot, sorry about that!! ok it’s fixed now :D

On Sep 12, 2022, at 4:23 PM, Simon Birrer ***@***.******@***.***>> wrote:



@sibirrer commented on this pull request.

Looks awesome!
One very minor last comment and then we can merge. Thank you very much!

________________________________

In tests/test_psfr.py<https://github.com/sibirrer/psfr/pull/7#discussion_r969010195>:

> @@ -280,3 +305,9 @@ def test_combine_psf():
     assert diff_input > diff_output

     npt.assert_raises(ValueError, combine_psf, kernel_list_input, kernel, stacking_option='BAD')
+
+
+
+
+
+test_fit_centroid_pso()


you don't need to execute the definition here. py.test does it anyways :)
This is my last request before approving :)

—
Reply to this email directly, view it on GitHub<https://github.com/sibirrer/psfr/pull/7#pullrequestreview-1104940408>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEA6XOTZCI3PKNHDQ7YA263V563OZANCNFSM6AAAAAAQJD3NSI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

--------------------------------------------
Anna M. Nierenberg

Assistant Professor
Department of Physics, UC Merced
annanierenberg.com<http://annanierenberg.com>
